components/site/src/sass/serde.rs

```diff
@@ -0,0 +1,477 @@
+/// instead of `{"key": "value"}` to express maps/dictionaries.
+#[derive(Default)]
+struct SassMapSerializer {
+    output: String,
+}
+
+#[derive(Debug)]
+struct SassMapSerializerError(String);
+
+type SassMapSerializerResult = std::result::Result<(), SassMapSerializerError>;
+
+impl serde::ser::Error for SassMapSerializerError {
+    fn custom<T: Display>(msg: T) -> Self {
+        SassMapSerializerError(msg.to_string())
+    }
+}
+
+impl Display for SassMapSerializerError {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_fmt(format_args!("SassMapSerializerError({})", self.0))
+    }
+}
+
+impl std::error::Error for SassMapSerializerError {}
+
+impl<'a> Serializer for &'a mut SassMapSerializer {
+    type Ok = ();
+    type Error = SassMapSerializerError;
+
+    type SerializeSeq = Self;
+    type SerializeTuple = Self;
+    type SerializeTupleStruct = Self;
+    type SerializeTupleVariant = Self;
+    type SerializeMap = Self;
+    type SerializeStruct = Self;
+    type SerializeStructVariant = Self;
+
+    fn serialize_bool(self, v: bool) -> std::result::Result<Self::Ok, Self::Error> {
+        self.output += if v { "true" } else { "false" };
+        Ok(())
+    }
+
+    fn serialize_i8(self, v: i8) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_i64(i64::from(v))
+    }
+
+    fn serialize_i16(self, v: i16) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_i64(i64::from(v))
+    }
+
+    fn serialize_i32(self, v: i32) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_i64(i64::from(v))
+    }
+
+    fn serialize_i64(self, v: i64) -> std::result::Result<Self::Ok, Self::Error> {
+        self.output += &v.to_string();
+        Ok(())
+    }
+
+    fn serialize_u8(self, v: u8) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_u64(u64::from(v))
+    }
+
+    fn serialize_u16(self, v: u16) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_u64(u64::from(v))
+    }
+
+    fn serialize_u32(self, v: u32) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_u64(u64::from(v))
+    }
+
+    fn serialize_u64(self, v: u64) -> std::result::Result<Self::Ok, Self::Error> {
+        self.output += &v.to_string();
+        Ok(())
+    }
+
+    fn serialize_f32(self, v: f32) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_f64(f64::from(v))
+    }
+
+    fn serialize_f64(self, v: f64) -> std::result::Result<Self::Ok, Self::Error> {
+        self.output += &v.to_string();
+        Ok(())
+    }
+
+    /// serialize any chars as if they were single-character strings
+    fn serialize_char(self, v: char) -> std::result::Result<Self::Ok, Self::Error> {
+        self.serialize_str(&v.to_string())
+    }
+
+    fn serialize_str(self, v: &str) -> std::result::Result<Self::Ok, Self::Error> {
+        self.output += "\"";
+        self.output += v;
```

Comment:
what if the string contains `"`?

Reply:
great point.  it seems like the most straightforward way to make this more robust is to first convert the `String` to a `serde_json::Value::String` and then write that out.  `Value::String` will do the right thing regarding escapes, which should function the same for a Sass literal as for JSON.

the downside here is obviously taking a dependency on the `serde_json` crate.  the alternative would be to simply check for the presence of characters that would require an escape and fail with a helpful error message.  that also seems acceptable, as I can't think of a compelling reason why someone would need quotes or other escape characters in this context.

@Keats do you have a recommendation on which route to go?

Reply:
can we do a naive escape (eg just a .replace)? It will do something weird if the user has an escaped string but that seems even less likely...

Reply:
yeah i like that.  i think what i'll do is naively escape double quotes and search for the characters that i know would need an escape, failing if i find any.  that seems like a good happy medium between allowing the might-happen thing and disallowing the shouldn't-happen thing with a helpful message.